src/inmanta/logging.py

```diff
@@ -0,0 +1,300 @@
+"""
+    Copyright 2023 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import logging
+import os
+import sys
+from argparse import Namespace
+from typing import Optional, TextIO
+
+import colorlog
+from colorlog.formatter import LogColors
+
+from inmanta import const
+from inmanta.stable_api import stable_api
+
+
+def _is_on_tty() -> bool:
+    return (hasattr(sys.stdout, "isatty") and sys.stdout.isatty()) or const.ENVIRON_FORCE_TTY in os.environ
+
+
+class Options(Namespace):
+    log_file: Optional[str]
+    log_file_level: Optional[str]
+    verbose: int
+    timed: bool = False
+
+
+"""
+This dictionary maps the Inmanta log levels to the corresponding Python log levels
+"""
+log_levels = {
+    "0": logging.ERROR,
+    "1": logging.WARNING,
+    "2": logging.INFO,
+    "3": logging.DEBUG,
+    "4": 2,
+    "ERROR": logging.ERROR,
+    "WARNING": logging.WARNING,
+    "INFO": logging.INFO,
+    "DEBUG": logging.DEBUG,
+    "TRACE": 2,
+}
+
+
+@stable_api
+class InmantaLoggerConfig:
+    """
+    A class that provides logging functionality for Inmanta projects.
+
+    Usage:
+    To use this class, you first need to call the `get_instance`. This method takes a `stream` argument
+    that specifies where the log messages should be sent to. If no `stream` is provided,
+    the log messages will be sent to standard output.
+
+    You can then call the `apply_options` method to configure the logging options. This method takes an `options`
+    argument that should be an Option object with the following attributes:
+    - `log_file`: if this attribute is set, the logs will be written to the specified file instead of the stream
+      specified in `create_default_handler`.
+    - `log_file_level`: the logging level for the file handler (if `log_file` is set).
+    - `verbose`: the verbosity level of the log messages.
+    - `timed`: if true,  adds the time to the formatter in the log lines.
+
+    The setup is not done in one step as we want logs for the cmd_parser, which will provide the options needed to configure
+    the 'final' logger with `apply_options`.
+
+    for more fine-grained configuration the following functions can be used as well:
+        - `set_log_level`
+        - `set_log_formatter`
+        - `set_logfile_location`
+    """
+
+    _instance: Optional["InmantaLoggerConfig"] = None
+
+    def __init__(self, stream: TextIO = sys.stdout) -> None:
```

Comment:
Just a question: Do we have a use case for passing something else than `sys.stdout` here?

Reply:
I saw here that Guillaume use stderr once:
https://code.inmanta.com/solutions/tools/yang-module-generator/-/blob/master/src/yang_module_generator/cli.py#L77-97

Reply:
I see. Thanks!